llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h

```diff
@@ -18,13 +18,12 @@
 namespace llvm {
 namespace SCEVPatternMatch {
 
-template <typename Val, typename Pattern>
-bool match(const SCEV *S, const Pattern &P) {
-  return P.match(S);
+template <typename Pattern> bool match(const SCEV *S, const Pattern &P) {
+  return const_cast<Pattern &>(P).match(S);
```

Comment:
Why do we need the const_cast?

Reply:
It's actually required because PatternMatch.h hasn't marked match functions `const`, and ScalarEvolution.cpp uses PatternMatch with SCEVPatternMatch:

```cpp
 10590 |   if (match(Step, m_CombineOr(m_scev_One(), m_scev_AllOnes()))) {
```

I can remove this in a follow-up, by marking match functions `const` in PatternMatch?

Reply:
I've fixed all the issues in this PR. Let me know if you prefer a separate PR, as I was originally planning.